test-suite/src/test/scala/org/scalajs/testsuite/niobuffer/BufferFactory.scala

```diff
@@ -0,0 +1,253 @@
+package org.scalajs.testsuite.niobuffer
+
+import scala.language.implicitConversions
+
+import scala.reflect._
+
+import java.nio._
+
+import scala.scalajs.js
+
+abstract class BufferFactory {
+  type BufferType <: Buffer with Comparable[BufferType]
+  type ElementType
+
+  implicit val elemClassTag: ClassTag[ElementType]
+
+  implicit def elemFromInt(value: Int): ElementType
+
+  implicit def elemToJSAny(elem: ElementType): js.Any
+
+  implicit def bufferAdapter(
+      buffer: BufferType): BufferAdapter[BufferType, ElementType]
+
+  val createsReadOnly: Boolean = false
+
+  def allocBuffer(capacity: Int): BufferType
+
+  def allocBuffer(pos: Int, limit: Int, capacity: Int): BufferType = {
+    val buf = allocBuffer(capacity)
+    buf.limit(limit).position(pos)
+    buf
+  }
+
+  def elemRange(start: Int, end: Int): Array[ElementType] =
+    (start until end).map(elemFromInt).toArray
+
+  def withContent(capacity: Int, content: ElementType*): BufferType =
+    withContent(0, capacity, capacity, content: _*)
+
+  def withContent(pos: Int, limit: Int, capacity: Int,
+      content: ElementType*): BufferType = {
+    val buf = allocBuffer(pos, limit, capacity)
+    buf.put(content.toArray)
+    buf.position(pos)
+    buf
+  }
+}
+
+object BufferFactory {
+  abstract class ByteBufferFactory extends BufferFactory {
```

Comment:
Because I refer to this class from the tests of other classes. The factories that create views of byte buffers take a `ByteBufferFactory` in their constructors. 

For symmetry, the other abstract classes are also defined here.